crates/cdk/src/wallet/proofs.rs

```diff
@@ -166,112 +167,389 @@ impl Wallet {
+            let selected_total = Amount::try_sum(selected_proofs.iter().map(|p| p.amount))?;
+            let fees = if opts.include_fees {
+                self.get_proofs_fee(&selected_proofs).await?
+            } else {
+                Amount::ZERO
             };
 
-            if proof_amount >= remaining_amount + fees {
-                remaining_amount = Amount::ZERO;
-                break;
+            if selected_total >= amount + fees {
+                return Ok(selected_proofs);
             }
+        }
 
-            remaining_amount = amount.checked_add(fees).ok_or(Error::AmountOverflow)?
-                - Amount::try_sum(selected_proofs.iter().map(|p| p.amount))?;
-            (proofs_larger, proofs_smaller) = proofs_smaller
-                .into_iter()
-                .skip(1)
-                .partition(|p| p.amount > remaining_amount);
+        Err(Error::InsufficientFunds)
+    }
+}
+
+fn sort_proofs(proofs: &mut Proofs, method: ProofSelectionMethod, amount: Amount) {
+    match method {
+        // Least fallback to largest
+        ProofSelectionMethod::Largest | ProofSelectionMethod::Least => {
+            proofs.sort_by(|a: &Proof, b: &Proof| b.cmp(a))
         }
+        ProofSelectionMethod::Closest => proofs.sort_by_key(|p| {
+            if p.amount > amount {
+                p.amount - amount
+            } else {
+                amount - p.amount
+            }
+        }),
+        ProofSelectionMethod::Smallest => proofs.sort(),
+    }
+}
 
-        if remaining_amount > Amount::ZERO {
-            if let Some(next_bigger) = next_bigger_proof {
-                return Ok(vec![next_bigger.clone()]);
+fn select_least_proofs_over_amount(
+    proofs: &Proofs,
+    amount: Amount,
+    fees: HashMap<Id, u64>,
+) -> Option<Vec<Proof>> {
+    tracing::trace!(
+        "Selecting LEAST proofs over amount {} with fees {:?}",
+        amount,
+        fees
+    );
+    let max_sum = Amount::try_sum(proofs.iter().map(|p| p.amount))
+        .ok()?
+        .checked_add(1.into())?;
+    if max_sum < amount || proofs.is_empty() || amount == Amount::ZERO {
+        return None;
+    }
+    let table_len = Into::<u64>::into(max_sum + 1.into()) as usize;
+    let mut dp = vec![None; table_len];
+    let mut paths = vec![Vec::<Proof>::new(); table_len];
+
+    dp[0] = Some(Amount::ZERO);
+
+    // Fill DP table and track paths
+    for proof in proofs {
+        let max_other_amounts: u64 = (max_sum - proof.amount).into();
+        for t in (0..=max_other_amounts).rev() {
+            if let Some(current_sum) = dp[t as usize] {
+                let new_sum = current_sum + proof.amount;
+                let target_index = (t + Into::<u64>::into(proof.amount)) as usize;
+
+                // If this sum has not been reached yet, or if the new sum is smaller, or if the new path is shorter
+                if dp[target_index].is_none()
+                    || dp[target_index].unwrap() > new_sum
```

Comment:
Do we know that this will always unwrap without panicking? If so can we change it to an `.expect()` with a reason why it wont panic. 

In general unwrap should only be used in tests in places where we're conformable with `unwrap` behavior because its checked in some other way lets use `expect`so we know why its safe to use.

Reply:
Yes, because the `None` check is right before it. I'll change to expect though.